src/jwt_authorizer/verify.py

```diff
@@ -0,0 +1,307 @@
+"""Verify a JWT."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from urllib.parse import urljoin
+
+import jwt
+from cachetools import TTLCache
+from cachetools.keys import hashkey
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat
+
+from jwt_authorizer.config import ALGORITHM
+from jwt_authorizer.util import base64_to_number
+
+if TYPE_CHECKING:
+    from aiohttp import ClientResponse, ClientSession, web
+    from logging import Logger
+    from jwt_authorizer.config import Config, Issuer
+    from typing import Any, Dict, List, Optional
+
+__all__ = [
+    "KeyClient",
+    "KeyClientException",
+    "TokenVerifier",
+    "UnknownAlgorithmException",
+    "UnknownKeyIdException",
+    "create_token_verifier",
+]
+
+
+class KeyClientException(Exception):
+    """Cannot retrieve the keys from an issuer."""
+
+
+class UnknownAlgorithmException(Exception):
+    """The issuer key was for an unsupported algorithm."""
+
+
+class UnknownKeyIdException(Exception):
+    """The reqeusted key ID was not found for an issuer."""
+
+
+class KeyClient:
+    """Client to retrieve a key from an issuer.
+
+    Handles retrieving the OpenID Connect metadata and the JWKS for an issuer
+    and extracting the keys.  Intended to be overridden by the test suite to
+    replace the get_url function.
+
+    Parameters
+    ----------
+    session : `aiohttp.ClientSession`
+        The session to use for making requests.
+    """
+
+    def __init__(self, session: ClientSession) -> None:
+        self.session = session
+
+    async def get_keys(self, issuer: Issuer) -> List[Dict[str, str]]:
+        """Fetch the key set for an issuer.
+
+        Parameters
+        ----------
+        url : `str`
+            URL to fetch.
+
+        Returns
+        -------
+        body : List[Dict[`str`, `str`]]
+            List of keys (in JWKS format) for the given issuer.
+
+        Raises
+        ------
+        KeyClientException
+            On failure to retrieve a set of keys from the issuer.
+        """
+        url = await self._get_jwks_uri(issuer)
+        if not url:
+            url = urljoin(issuer.url, ".well-known/jwks.json")
+
+        r = await self.get_url(url)
+        if r.status != 200:
+            msg = f"Cannot retrieve keys from {issuer.url}"
+            raise KeyClientException(msg)
+
+        body = await r.json()
+        if "keys" not in body:
+            msg = f"No keys property in JWKS metadata for {issuer.url}"
```

Comment:
Small nitpick, but it looks like you're discovering the url then defaulting to the issuer url, but you don't say what URL you're getting in the errors (you use issuer.url), might be helpful to emit the URL somewhere, but it will probably never happen.

Reply:
Yup, good call.